lib/stringify.js

```diff
@@ -308,50 +317,134 @@ function escapeFactory(options) {
+                            )
+                        )
+                    )
+                )
+            ) {
+                queue += BACKSLASH;
+            } else if (character === LINE) {
+                afterNewLine = true;
+            } else if (afterNewLine) {
+                if (
+                    character === ANGLE_BRACKET_CLOSE ||
+                    character === DASH
+                ) {
+                    queue += BACKSLASH;
+                } else if (
+                    character !== SPACE &&
+                    character !== TAB &&
+                    character !== CARRIAGE &&
+                    character !== VERTICAL_TAB &&
+                    character !== FORM_FEED
+                ) {
+                    afterNewLine = false
+                }
             }
 
-            if (value[0] == '(' && previous &&
-                previous.type == 'linkReference' &&
-                previous.referenceType == 'shortcut')
-            {
-                escaped = '\\' + escaped;
+            queue += character;
+        }
+
+        escaped = queue;
+
+        /*
+         * Multi-node versions.
+         */
+
+        if (siblings && node.type == 'text') {
+            index = siblings.indexOf(node);
+            prev = siblings[index - 1];
+
+            /*
+             * Check for an opening parentheses after a
+             * link-reference (which can be joined by
+             * white-space).
+             */
+
+            if (
+                prev &&
+                prev.type == 'linkReference' &&
+                prev.referenceType == 'shortcut'
+            ) {
+                position = -1;
+                length = escaped.length;
+
+                while (++position < length) {
+                    character = escaped.charAt(position);
+
+                    if (character === SPACE || character === TAB) {
+                        continue;
+                    }
+
+                    if (character === PARENTHESIS_OPEN) {
+                        escaped = escaped.slice(0, position) +
+                            BACKSLASH + escaped.slice(position);
+                    }
+
+                    break;
+                }
             }
 
-            if (value[0] == ':' && previous.type == 'text') {
-                // Unlike previous cases, the context for this pattern can be
-                // split across more than a single text node.
-                var before = '', i = index - 1;
-                do {
-                    before = parent.children[i].value + before;
-                } while (
-                    before.length < 5 &&
-                    --i >= 0 &&
-                    parent.children[i].type == 'text'
-                )
+            /*
+             * Ensure non-auto-links are not seen links.
+             * This pattern needs to check the preceding
+             * nodes too.
+             */
+
+            if (prev && prev.type == 'text' && value.charAt(0) == COLON) {
```

Comment:
Shouldn’t this be `commonmark` only (or even anything non-GFM)? Just like the one-node method above?